File: server/session.go
//go:build !js

package server

import "sync"

type sessionIDStore struct {
	sessions sync.Map
}

func newSessionIDStore() SessionIDStore {
	return &sessionIDStore{}
}

func (s *sessionIDStore) Store(sessionID string) {
	s.sessions.Store(sessionID, true)
}

func (s *sessionIDStore) Range(f func(sessionID string) bool) {
	s.sessions.Range(func(sessionID, value any) bool {
		return f(sessionID.(string))
	})
}

func (s *sessionIDStore) Load(sessionID string) bool {
	_, ok := s.sessions.Load(sessionID)
	return ok
}

func (s *sessionIDStore) Delete(sessionID string) {
	s.sessions.Delete(sessionID)
}

File: worker/session.ts
import { DurableObject } from "cloudflare:workers";

type SSESession = {
  sessionId: string;
};

type Env = unknown;

export class SessionStore extends DurableObject {
  sql: SqlStorage;
  constructor(ctx: DurableObjectState, env: Env) {
    super(ctx, env);
    this.sql = ctx.storage.sql;

    this.sql.exec(`CREATE TABLE IF NOT EXISTS sessions(
      session_id    TEXT PRIMARY KEY
    );`);
  }

  async store(sessionId: string, session: SSESession): Promise<void> {
    this.sql.exec("INSERT INTO sessions (session_id) VALUES ($1)", [sessionId]);
  }

  async range(
    f: (sessionId: string, session: SSESession) => boolean
  ): Promise<void> {
    const rows = await this.sql.exec("SELECT * FROM sessions");
    for (const row of rows) {
      const session: SSESession = { sessionId: row.session_id };
      if (!f(row.session_id, session)) {
        break;
      }
    }
  }

  async load(sessionId: string): Promise<SSESession | null> {
    const row = await this.sql.exec(
      "SELECT * FROM sessions WHERE session_id = $1",
      [sessionId]
    );
    if (row.length === 0) {
      return null;
    }
    return { sessionId: row[0].session_id };
  }

  async delete(sessionId: string): Promise<void> {
    this.sql.exec("DELETE FROM sessions WHERE session_id = $1", [sessionId]);
  }
}
